README.md
# xchain_dart

Lightweight Dart library for wallet apps.

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 1.0.2

- Validating testnet address on read-only client.
- Fixed issue on retrieving transaction history of coinbase transactions.
- Added missing hash field.

## 1.0.1

- Implemented validateAddress function.
- Removed optional network argument in XChainClient constructor.
- Changed network type from String to NetworkType.

## 1.0.0

- Bitcoin client.

## 0.0.1

- Base chain client.


test/xchain_crypto/utils_test.dart
import 'package:test/test.dart';
import 'package:xchain_dart/src/xchain_crypto/utils.dart';

void main() {
  test('is empty', () {
    String source = '';
    List<AssetAddress> addresses = [];
    try {
      addresses = substractAddress(source);
    } catch (err) {
      expect((err as ArgumentError).message, 'Input is empty');
    } finally {
      expect(addresses, isEmpty);
    }
  });

  test('contains spaces', () {
    String source =
        '3QaesQ25kJc4tyCQM5wJ54ky39DNsUMx7Z 0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d';
    try {
      substractAddress(source);
    } catch (err) {
      expect((err as ArgumentError).message, 'Illegal character');
    }
  });

  test('starts with bitcoin chain prefix', () {
    String source = 'bitcoin:3QaesQ25kJc4tyCQM5wJ54ky39DNsUMx7Z';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 1);
    expect(addresses.first.address, '3QaesQ25kJc4tyCQM5wJ54ky39DNsUMx7Z');
    expect(addresses.first.asset, 'BTC.BTC');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('starts with bitcoin chain prefix', () {
    String source = 'BITCOIN:BC1QFPFPW7DLQT3AZCYQXMCUZM5L2XRM6PD6ZVSE9Q';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 1);
    expect(
        addresses.first.address, 'BC1QFPFPW7DLQT3AZCYQXMCUZM5L2XRM6PD6ZVSE9Q');
    expect(addresses.first.asset, 'BTC.BTC');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('starts with bitcoin cash chain prefix', () {
    String source = 'bitcoincash:qpl4lfjq7emfg8p4akr6p27dap5duj35zcc82aqul5';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 1);
    expect(
        addresses.first.address, 'qpl4lfjq7emfg8p4akr6p27dap5duj35zcc82aqul5');
    expect(addresses.first.asset, 'BCH.BCH');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('starts with ethereum chain prefix', () {
    String source = 'ethereum:0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d';
    List<AssetAddress> addresses = substractAddress(source);
    expect(
        addresses.first.address, '0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d');
    expect(addresses.first.asset, 'ETH.ETH');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('starts with dragonsdex chain prefix', () {
    String source = 'dragonsdex:0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d';
    List<AssetAddress> addresses = [];
    try {
      addresses = substractAddress(source);
    } catch (err) {
      expect((err as ArgumentError).message, 'Unsupported chain prefix');
    } finally {
      expect(addresses.length, 0);
    }
  });

  test('starts with binance chain prefix', () {
    String source = 'binance:bnb1vxyxxkqdke8r55r6fzhprtj8qwgecudj0h5svr';
    List<AssetAddress> addresses = substractAddress(source);
    expect(
        addresses.first.address, 'bnb1vxyxxkqdke8r55r6fzhprtj8qwgecudj0h5svr');
    expect(addresses.first.asset, 'BNB.BNB');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('bitcoin legacy address without chain prefix', () {
    String source = '16we9adsewmBDKv5CSgeRMZPo3RadcgVZV';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 2);
    expect(addresses.first.address, '16we9adsewmBDKv5CSgeRMZPo3RadcgVZV');
    expect(addresses[0].asset, 'BCH.BCH');
    expect(addresses[1].asset, 'BTC.BTC');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('bitcoin segwit address without chain prefix', () {
    String source = '3QaesQ25kJc4tyCQM5wJ54ky39DNsUMx7Z';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 2);
    expect(addresses.first.address, '3QaesQ25kJc4tyCQM5wJ54ky39DNsUMx7Z');
    expect(addresses[0].asset, 'BTC.BTC');
    expect(addresses[1].asset, 'LTC.LTC');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('bitcoin native segwit address without chain prefix', () {
    String source = 'bc1qfw00pnu77vvw3r8fpterjukx0u3nj26n724pq3';
    List<AssetAddress> addresses = substractAddress(source);
    expect(addresses.length, 1);
    expect(
        addresses.first.address, 'bc1qfw00pnu77vvw3r8fpterjukx0u3nj26n724pq3');
    expect(addresses.first.asset, 'BTC.BTC');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('ethereum address without chain prefix', () {
    String source = '0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d';
    List<AssetAddress> addresses = substractAddress(source);
    expect(
        addresses.first.address, '0xC52A857FDa38994CB6CC8e0DE2AEDD67a7353e0d');
    expect(addresses.first.asset, 'ETH.ETH');
    expect(addresses.first.networkType, 'mainnet');
  });

  test('unsupported Chain address', () {
    String source = 'AJXPYa2aYizxXfhcEcmom1xuEyZLF6DX5b';
    List<AssetAddress> addresses = [];
    try {
      addresses = substractAddress(source);
    } catch (err) {
      expect((err as ArgumentError).message, 'Unsupported chain');
    } finally {
      expect(addresses, isEmpty);
    }
  });
}


test/xchain_crypto/crypto_test.dart
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:xchain_dart/src/xchain_crypto/crypto.dart';

void main() {
  group('generate mnemonic (BIP39)', () {
    test('generate default mnemonic', () {
      var mnemonic = generateMnemonic();
      var words = mnemonic.split(' ');

      expect(words.length, 12);
    });

    test('generate 12-word mnemonic', () {
      var mnemonic = generateMnemonic(size: 12);
      var words = mnemonic.split(' ');

      expect(words.length, 12);
    });

    test('generate 24-word mnemonic', () {
      var mnemonic = generateMnemonic(size: 24);
      var words = mnemonic.split(' ');

      expect(words.length, 24);
    });

    test('generate neither a 12 or 24-word mnemonic', () {
      var mnemonic = generateMnemonic(size: 8);
      var words = mnemonic.split(' ');

      expect(words.length, 12);
    });
  });

  group('validate mnemonic', () {
    test('validate 12-word mnemonic', () {
      var mnemonic = generateMnemonic(size: 12);
      bool correctMnemonic = validateMnemonic(mnemonic);

      expect(correctMnemonic, true);
    });

    test('validate 24-word mnemonic', () {
      var mnemonic = generateMnemonic(size: 24);
      bool correctMnemonic = validateMnemonic(mnemonic);

      expect(correctMnemonic, true);
    });

    test('validate invalid mnemonic', () {
      const mnemonic =
          'flush viable fury sword mention dignity ethics secret nasty gallery teach wrong';
      bool incorrectMnemonic = validateMnemonic(mnemonic);

      expect(incorrectMnemonic, false);
    });

    test('validate incomplete mnemonic', () {
      const mnemonic = 'neck grocery crumble super myself license ghost';
      bool incorrectMnemonic = validateMnemonic(mnemonic);

      expect(incorrectMnemonic, false);
    });
  });

  group('generate seed', () {
    test('generate seed hex from mnemonic', () {
      String seed = getSeedHex(
          "update elbow source spin squeeze horror world become oak assist bomb nuclear");

      expect(seed,
          '77e6a9b1236d6b53eaa64e2727b5808a55ce09eb899e1938ed55ef5d4f8153170a2c8f4674eb94ce58be7b75922e48e6e56582d806253bd3d72f4b3d896738a4');
    });

    test('generate seed hex from mnemonic and passphrase', () {
      String seed = getSeedHex(
          "update elbow source spin squeeze horror world become oak assist bomb nuclear",
          passphrase: "xchain rocks");

      expect(seed,
          '34f9182bda6d4cb579f66a932dc0941bf990ade04d41ac1e55801e70857ed253af6cf5a9fe76af56f18ba01a07aa2ac3ad0115264cbd365f391684acfc29cae5');
    });

    // test('generate seed hex from invalid mnemonic', () {
    //   String seed = getSeedHex(
    //       "update elbow source spin squeeze horror world become oak assist bomb wrong");
    //
    //   expect(seed, isArgumentError);
    // });

    test('generate seed from mnemonic', () {
      Uint8List seed = getSeed('basket actual');
      // Uint8List seed = getSeed(
      //     'update elbow source spin squeeze horror world become oak assist bomb nuclear');
      expect(seed, [
        92,
        242,
        212,
        168,
        176,
        53,
        94,
        144,
        41,
        91,
        223,
        197,
        101,
        160,
        34,
        164,
        9,
        175,
        6,
        61,
        83,
        101,
        187,
        87,
        191,
        116,
        217,
        82,
        143,
        73,
        75,
        250,
        68,
        0,
        245,
        61,
        131,
        73,
        184,
        15,
        218,
        228,
        64,
        130,
        215,
        249,
        84,
        30,
        29,
        186,
        43,
        0,
        59,
        207,
        236,
        157,
        13,
        83,
        120,
        28,
        166,
        118,
        101,
        31
      ]);
    });
  });
}


test/xchain_bitcoin/bitcoin_client_test.dart
import 'package:bitcoin_dart/bitcoin_flutter.dart';
import 'package:test/test.dart';
import 'package:xchain_dart/xchaindart.dart';

void main() {
  // https://iancoleman.io/bip39/
  const phrase =
      'canyon throw labor waste awful century ugly they found post source draft';

  // m/44'/0'/0'/0/0
  const addrPath0 = '12tSpVdC9CAwod9CFaw33JL9o7JngpE2pJ';
  // m/84'/1'/0'/0/0
  const addrPath1 = 'tb1q669kqq0ykrzgx337w3sj0kdf6zcuznvff34z85';
  // m/84'/0'/0'/0/0
  const addrPath2 = 'bc1qan023kcugy7gslksnffh8ej5rhqajjvu78r3tk';
  // m/84'/0'/0'/0/1
  const addrPath3 = 'bc1qnlnwpt3e3ur0k2cwppypqsr4jfn8pdp3ty0z04';
  // Satoshis address
  const addrPath4 = '1HLoD9E4SDFFPDiYfNYnkBLQ85Y51J3Zb1';

  group('config and setup', () {
    XChainClient client = BitcoinClient(
      phrase,
    );

    test('check if the readOnlyClient flag is set', () {
      expect(client.readOnlyClient, false);
    });
    test('get default network', () {
      var networkType = client.getNetwork();
      expect(networkType, bitcoin);
    });
    test('set testnet network', () {
      client.setNetwork(testnet);
      var networkType = client.getNetwork();
      expect(networkType, testnet);
      expect(client.address, addrPath1);
    });
    test('set phrase', () {
      client.setNetwork(testnet);
      String address = client.setPhrase(phrase, 0);
      expect(address, addrPath1);
    });
  });
  group('bitcoin-client', () {
    XChainClient client = new BitcoinClient(phrase);
    test('check valid address on creation', () {
      expect(client.address, addrPath2);
    });
    test('check valid address on creation', () {
      expect(client.getAddress(1), addrPath3);
    });
  });

  group('empty bitcoin-lite-client', () {
    XChainClient client = new BitcoinClient.readonly(addrPath0);
    test('check if the readOnlyClient flag is set', () {
      expect(client.readOnlyClient, true);
    });
    test('check if address is set on creation', () {
      expect(client.address, addrPath0);
    });
    test('check balance', () async {
      List balances = await client.getBalance(addrPath0, 'BTC.BTC');
      expect(balances.length, 1);
      expect(balances.first['amount'], 0.0);
    });
  });

  group('non-empty bitcoin-lite-client', () {
    XChainClient client = new BitcoinClient.readonly(addrPath4);
    test('check if the readOnlyClient flag is set', () {
      expect(client.readOnlyClient, true);
    });
    test('check if address is set on creation', () {
      expect(client.address, addrPath4);
    });
    test('check balance', () async {
      List balances = await client.getBalance(addrPath4, 'BTC.BTC');
      expect(balances.length, equals(1));
      expect(balances.first['amount'], greaterThan(0.01597));
    });
  });

  group('transaction history', () {
    XChainClient client = new BitcoinClient.readonly(addrPath4);
    test('get specific tx history', () async {
      Map txData = await client.getTransactionData(
          'b12dd481c49c01c3570672e2a5f72efb2deb74a10a5d27a9cbe4483160fe9565');
      expect(
          txData.containsValue(
              '000000000000000000008d1e5a3c919bcd0db96ce149b88da6f6246b0dab3f12'),
          true);
    });
    test('get all tx history', () async {
      List transactions =
          await client.getTransactions('35gHuxDkYMEETK5bXCTDen9rCv5dGz3i7Z', 3);
      Map tx = transactions.first;
      bool asset = tx.containsValue('BTC.BTC');
      expect(asset, true);
      expect(transactions.length, 3);
    });
    test('get all tx history from miner on testnet', () async {
      client.setNetwork(testnet);
      List transactions = await client.getTransactions(
          'tb1qsgx55dp6gn53tsmyjjv4c2ye403hgxynxs0dnm', 1);
      Map tx = transactions.first;
      bool asset = tx.containsValue('BTC.tBTC');
      expect(asset, true);
      expect(transactions.length, 1);
    });
  });

  group('check for address validity', () {
    XChainClient client = new BitcoinClient.readonly(addrPath4);
    test('check default address', () async {
      expect(client.validateAddress(client.address), true);
    });
    test('check native SegWit (P2WPKH) address', () async {
      expect(client.validateAddress(addrPath2), true);
    });
    test('check native SegWit (P2WPKH) testnet address', () async {
      client.setNetwork(testnet);
      expect(client.validateAddress(addrPath1), true);
    });
  });
}


lib/src/xchain_bitcoin/bitcoin_client.dart
import 'dart:convert';

import 'package:bip32/bip32.dart' as bip32;
import 'package:bip39/bip39.dart' as bip39;
import 'package:bitcoin_dart/bitcoin_flutter.dart';

import '../../xchaindart.dart';

class BitcoinClient implements XChainClient {
  @override
  late String address;

  @override
  NetworkType network = bitcoin;

  @override
  late bool readOnlyClient;

  @override
  late String seed;

  NetworkHelper _networkHelper = NetworkHelper();

  static const _denominator = 100000000;

  BitcoinClient(this.seed) {
    readOnlyClient = false;
    int walletIndex = 0;
    address = getAddress(walletIndex);
  }

  BitcoinClient.readonly(this.address) {
    readOnlyClient = true;
    address = this.address;
  }

  @override
  getAddress(walletIndex) {
    if (walletIndex < 0) {
      throw ('index must be greater than zero');
    }

    final seedUint8List = bip39.mnemonicToSeed(seed);
    final root = bip32.BIP32.fromSeed(seedUint8List);

    final address;

    // BIP84 (BIP44 for native segwit)
    if (network == testnet) {
      final node = root.derivePath("m/84'/1'/0'/0/$walletIndex");
      address = P2WPKH(
              data: new PaymentData(pubkey: node.publicKey), network: testnet)
          .data
          .address;
    } else {
      final node = root.derivePath("m/84'/0'/0'/0/$walletIndex");
      address =
          P2WPKH(data: new PaymentData(pubkey: node.publicKey)).data.address;
    }

    return address;
  }

  @override
  getBalance(address, assets) async {
    List balances = [];

    String uri = '${getExplorerAddressUrl(address)}';
    String responseBody = await _networkHelper.getData(uri);
    num funded = jsonDecode(responseBody)['chain_stats']['funded_txo_sum'];
    num spend = jsonDecode(responseBody)['chain_stats']['spent_txo_sum'];
    num amount = (funded - spend) / _denominator;

    String _asset;
    network == testnet ? _asset = 'tBTC' : _asset = 'BTC';

    balances.add({
      'asset': 'BTC.$_asset',
      'amount': amount,
      'image': 'https://s2.coinmarketcap.com/static/img/coins/64x64/1.png'
    });

    return balances;
  }

  @override
  getExplorerAddressUrl(address) {
    return '${this.getExplorerUrl()}/address/${address}';
  }

  @override
  getExplorerTransactionUrl(txId) {
    return '${this.getExplorerUrl()}/tx/${txId}';
  }

  @override
  getExplorerUrl() {
    if (network == bitcoin) {
      return 'https://blockstream.info/api';
    } else if (network == testnet) {
      return 'https://blockstream.info/testnet/api';
    } else {
      throw ArgumentError('Unsupported network');
    }
  }

  @override
  getFees(params) {
    List fees = [
      {
        "type": "byte",
        "fastest": 300,
        "fast": 275,
        "average": 250,
      }
    ];
    return fees;
  }

  @override
  getNetwork() {
    return network;
  }

  @override
  getTransactionData(txId) async {
    var _txData = {};

    String _uri = '${getExplorerTransactionUrl(txId)}';
    String _responseBody = await _networkHelper.getData(_uri);
    var _rawTx = jsonDecode(_responseBody);

    var _confirmed = _rawTx['status']['confirmed'];
    var _hash = _rawTx['status']['block_hash'];
    var _date = DateTime.now();
    if (_confirmed == true) {
      var epoch = _rawTx['status']['block_time'];
      _date =
          new DateTime.fromMillisecondsSinceEpoch(epoch * 1000, isUtc: false);
    }

    List<Map> _from = [];
    _rawTx['vin'].forEach((tx) {
      Map _txMap = tx;
      _txMap.forEach((key, value) {
        if (key == 'prevout') {
          Map _prevoutMap = value;
          late String _address;
          late double _amount;
          _prevoutMap.forEach((subkey, subvalue) {
            if (subkey == 'scriptpubkey_address') {
              _address = subvalue;
            }
            if (subkey == 'value') {
              _amount = subvalue / _denominator;
            }
          });
          if (_address.isNotEmpty) {
            var map = {'address': _address, 'amount': _amount};
            _from.add(map);
          }
        }
      });
    });

    List<Map> _to = [];
    _rawTx['vout'].forEach((tx) {
      Map _txMap = tx;
      late String _address;
      late double _amount;
      _txMap.forEach((key, value) {
        if (key == 'scriptpubkey_address') {
          _address = value;
        }

        if (key == 'value') {
          _amount = value / _denominator;
        }
      });
      if (_address.isNotEmpty) {
        var map = {'address': _address, 'amount': _amount};
        _to.add(map);
      }
    });

    String _asset;
    network == testnet ? _asset = 'tBTC' : _asset = 'BTC';

    if (_rawTx != null) {
      _txData.addAll({
        'asset': 'BTC.$_asset',
        'from': _from,
        'to': _to,
        'date': _date,
        'type': "transfer",
        'hash': _hash,
        'confirmed': _confirmed,
      });
    }
    return _txData;
  }

  @override
  getTransactions(address, [limit]) async {
    String _addressUri = '${getExplorerAddressUrl(address)}';
    String _addrResponseBody = await _networkHelper.getData(_addressUri);
    var _rawAddressStats = jsonDecode(_addrResponseBody);

    // Retrieve the number of transactions for an address.
    int _txCount = _rawAddressStats['chain_stats']['tx_count'] ?? 0;

    // Blockstream api limits tx results to 25 per page.
    int _pages = (_txCount / 25).ceil();

    // Avoid retrieving more data then explicitly requested.
    if (limit != null) {
      int _limit = (limit / 25).ceil();
      if (_limit < _pages) {
        _pages = _limit;
      }
    }

    List _txData = [];
    String _lastTx = '';

    for (int i = 0; i < _pages; i++) {
      String _txUri = '$_addressUri/txs/chain/$_lastTx';
      String _txResponseBody = await _networkHelper.getData(_txUri);
      var _rawTxs = jsonDecode(_txResponseBody);
      _lastTx = _rawTxs.last['txid'];

      for (var _rawTx in _rawTxs) {
        var _confirmed = _rawTx['status']['confirmed'];
        String _txid = _rawTx['txid'];
        var _date = DateTime.now();
        if (_confirmed == true) {
          var _epoch = _rawTx['status']['block_time'];
          _date = new DateTime.fromMillisecondsSinceEpoch(_epoch * 1000,
              isUtc: false);
        }

        List<Map> _from = [];
        _rawTx['vin'].forEach((tx) {
          Map _txMap = tx;
          _txMap.forEach((key, value) {
            if (key == 'prevout') {
              Map _prevoutMap = value ?? {};
              if (_prevoutMap.isNotEmpty) {
                String _address = '';
                double _amount = 0.0;
                _prevoutMap.forEach((subkey, subvalue) {
                  if (subkey == 'scriptpubkey_address') {
                    _address = subvalue;
                  }
                  if (subkey == 'value') {
                    _amount = subvalue / _denominator;
                  }
                });
                if (_address.isNotEmpty) {
                  var _map = {'address': _address, 'amount': _amount};
                  _from.add(_map);
                }
              }
            }
          });
        });

        List<Map> _to = [];
        _rawTx['vout'].forEach((tx) {
          Map _txMap = tx;
          String _address = '';
          double _amount = 0.0;
          _txMap.forEach((key, value) {
            if (key == 'scriptpubkey_address') {
              _address = value;
            }
            if (key == 'value') {
              _amount = value / _denominator;
            }
          });
          var _map = {'address': _address, 'amount': _amount};
          _to.add(_map);
        });

        String _asset;
        network == testnet ? _asset = 'tBTC' : _asset = 'BTC';

        _txData.add({
          'asset': 'BTC.$_asset',
          'from': _from,
          'to': _to,
          'date': _date,
          'type': "transfer",
          'txid': _txid,
          'confirmed': _confirmed,
        });
      }
    }

    if (limit == null) {
      return _txData;
    } else {
      return _txData.sublist(0, limit);
    }
  }

  @override
  purgeClient() {
    // When a wallet is "locked" the private key should be purged in each client by setting it back to null.
  }

  @override
  setNetwork(newNetwork) {
    network = newNetwork;
    if (readOnlyClient == false) {
      address = getAddress(0);
    }
  }

  @override
  setPhrase(mnemonic, walletIndex) {
    seed = mnemonic;
    address = getAddress(walletIndex);
    return address;
  }

  @override
  transfer(params) {
    String txHash =
        '59bbb95bbe740ad6acf24509d38f13f83ca49d6f11207f6a162999ffc5863b77';
    return txHash;
  }

  @override
  validateAddress(address) {
    bool result = Address.validateAddress(address, network);
    return result;
  }
}


lib/src/xchain_crypto/utils.dart
import 'package:http/http.dart' as http;

// Create address from the public key.
substractAddress(String source) {
  String address;

  // start with empty list
  List<AssetAddress> addresses = [];

  // If source is empty or null
  if (source.isEmpty) {
    throw ArgumentError('Input is empty');
  }

  // matches on spaces
  RegExp regExpSpaces = new RegExp(r' ');
  bool hasIllegalChars = regExpSpaces.hasMatch(source);
  if (hasIllegalChars) {
    throw ArgumentError('Illegal character');
  }

  // starts with chain prefix
  RegExp regExpPrefix = new RegExp(r':');
  bool hasPrefix = regExpPrefix.hasMatch(source);

  if (hasPrefix) {
    RegExp regex2 = RegExp(r'^(.+):(.+)');
    var matches2 = regex2.firstMatch(source);
    String prefix = matches2!.group(1)!.toLowerCase();
    if (prefix == 'bitcoin' ||
        prefix == 'bitcoincash' ||
        prefix == 'binance' ||
        prefix == 'ethereum') {
      address = matches2.group(2)!;
      // identify chain and assets
      addresses = _identifyChain(address, prefix);
    } else {
      throw ArgumentError('Unsupported chain prefix');
    }
  } else {
    address = source;
    // identify chain and assets
    addresses = _identifyChain(address);
  }

  return addresses;
}

_identifyChain(String address, [String? prefix]) {
  List<AssetAddress> _addresses = [];

  if (prefix == 'bitcoin') {
    _addresses.add(AssetAddress(address, 'BTC.BTC', 'mainnet'));
  } else if (prefix == 'bitcoincash') {
    _addresses.add(AssetAddress(address, 'BCH.BCH', 'mainnet'));
  } else if (prefix == 'ethereum') {
    _addresses.add(AssetAddress(address, 'ETH.ETH', 'mainnet'));
  } else if (prefix == 'binance') {
    _addresses.add(AssetAddress(address, 'BNB.BNB', 'mainnet'));
  }

  // Binance address starts with bnb and has 42 characters
  else if (address.startsWith(new RegExp(r'(^bnb[A-z,0-9]{39})'))) {
    _addresses.add(AssetAddress(address, 'BNB.BNB', 'mainnet'));
  }
  // Binance testnet address starts with tbnb and has 42 characters
  else if (address.startsWith(new RegExp(r'(^tbnb[A-z,0-9]{39})'))) {
    _addresses.add(AssetAddress(address, 'BNB.BNB', 'testnet'));
  }
  // Bitcoin Legacy address starts with 1 and has 34 or less characters
  else if (address.startsWith(new RegExp(r'(^1[A-z,0-9]{33})'))) {
    _addresses.add(AssetAddress(address, 'BCH.BCH', 'mainnet'));
    _addresses.add(AssetAddress(address, 'BTC.BTC', 'mainnet'));
  }
  // Bitcoin & Litecoin Segwit address starts with 3 and has 34 characters
  else if (address.startsWith(new RegExp(r'(^3[A-z,0-9]{33})'))) {
    _addresses.add(AssetAddress(address, 'BTC.BTC', 'mainnet'));
    _addresses.add(AssetAddress(address, 'LTC.LTC', 'mainnet'));
  }
  // Bitcoin Native-Segwit address starts with bc1 and has 42 characters
  else if (address.startsWith(new RegExp(r'(^bc1[A-z,0-9]{39})'))) {
    _addresses.add(AssetAddress(address, 'BTC.BTC', 'mainnet'));
  }
  // Bitcoin Cash address starts with q and has 42 characters
  else if (address.startsWith(new RegExp(r'(^q[A-z,0-9]{41})'))) {
    _addresses.add(AssetAddress(address, 'BCH.BCH', 'mainnet'));
  }
  // Litecoin address starts with L and has 34 or less characters
  else if (address.startsWith(new RegExp(r'(^L[A-z,0-9]{33})'))) {
    _addresses.add(AssetAddress(address, 'LTC.LTC', 'mainnet'));
  }
  // Litecoin address starts with ltc and has 43 or less characters
  else if (address.startsWith(new RegExp(r'(^ltc[A-z,0-9]{40})'))) {
    _addresses.add(AssetAddress(address, 'LTC.LTC', 'mainnet'));
  }
  // Dogecoin address starts with D and has 34 or less characters
  else if (address.startsWith(new RegExp(r'(^D[A-z,0-9]{33})'))) {
    _addresses.add(AssetAddress(address, 'DOGE.DOGE', 'mainnet'));
  }
  // Ethereum address starts with 0x and has 42 characters
  else if (address.startsWith(new RegExp(r'(^0x[A-z,0-9]{40})'))) {
    _addresses.add(AssetAddress(address, 'ETH.ETH', 'mainnet'));
  }
  // Thorchain address starts with thor and has 43 characters
  else if (address.startsWith(new RegExp(r'(^thor[A-z,0-9]{39})'))) {
    _addresses.add(AssetAddress(address, 'RUNE.RUNE', 'mainnet'));
  }
  // Thorchain testnet address starts with tthor and has 44 characters
  else if (address.startsWith(new RegExp(r'(^tthor[A-z,0-9]{39})'))) {
    _addresses.add(AssetAddress(address, 'RUNE.RUNE', 'testnet'));
  } else {
    throw ArgumentError('Unsupported chain');
  }
  return _addresses;
}

class AssetAddress {
  String address;
  String asset;
  String networkType;

  AssetAddress(this.address, this.asset, this.networkType);
}

class NetworkHelper {
  Future getData(String uri, [Map<String, String>? header]) async {
    // print('uri: $uri');
    // print('header: $header');
    var url = Uri.parse(uri);
    http.Response response = await http.get(url, headers: header);
    if (response.statusCode == 200) {
      String data = response.body;
      return data;
    } else if (response.statusCode == 404) {
      String data = response.body;
      return data;
    } else {
      print(response.statusCode);
      throw 'Problem with the get request';
    }
  }
}
